### PR TITLE
Fix and document a bunch of bugs in the image updating regexes.

### PR DESCRIPTION
**Pull request by paulbellamy**
This was prompted by flux failing to release an image, because I had the `name` and `image` attributes of the map in a different order.

There are still a slew of bugs remaining in this (and policy annotations), but I think they are mostly documented now.

PRed into the git-sync branch because these files have all moved and changed.

### CODE REVIEW

File: cluster/kubernetes/policies.go [hunk 1]
```diff
@@ -89,21 +89,40 @@ func updateAnnotations(def []byte, f func(map[string]string) map[string]string)
...
+			} `yaml:"spec"`
+		} `yaml:"template"`
+	} `yaml:"spec"`
 }
 
-func parseAnnotations(def []byte) (map[string]string, error) {
-	var p annotationsParser
-	if err := yaml.Unmarshal(def, &p); err != nil {
-		return nil, errors.Wrap(err, "decoding annotations")
+func (m Metadata) AnnotationsOrNil() map[string]string {
```
Comment (squaremo) [inline cluster/kubernetes/policies.go]: There's already some code for parsing manifests in `cluster/kubernetes/resources`, could it be used here _mutatis mutandis_?

File: cluster/kubernetes/policies_test.go [hunk 2]
```diff
@@ -99,7 +99,7 @@ func TestUpdatePolicies(t *testing.T) {
 	} {
 		caseIn := templToString(t, annotationsTemplate, c.in)
 		caseOut := templToString(t, annotationsTemplate, c.out)
-		out, err := UpdatePolicies([]byte(caseIn), c.update)
+		out, err := (&Cluster{}).UpdatePolicies([]byte(caseIn), c.update)
```
Comment (squaremo) [inline cluster/kubernetes/policies_test.go]: Good evidence that we should split out `Cluster` and um, `ManifestRepo` (or something). But doesn't need to be done here and now.

File: cluster/kubernetes/update.go [hunk 3]
```diff
@@ -81,89 +81,95 @@ func updatePodController(def []byte, newImageID flux.ImageID) ([]byte, error) {
...
+	// Detect how indented the "containers" block is.
+	newDef := string(def)
+	matches := regexp.MustCompile(`( +)containers:.*`).FindStringSubmatch(newDef)
+	if len(matches) != 2 {
+		return fmt.Errorf("could not find container specs")
+	}
+	indent := matches[1]
+
+	// Replace the container images
+	containersRE := regexp.MustCompile(`(?m:^` + indent + `containers:\s*(?:#.*)*$(?:\n(?:` + indent + `[-\s].*)?)*)`)
```
Comment (squaremo) [inline cluster/kubernetes/update.go]: Some comments saying what these regexps are for (and more distinct names) would be good here.

(Yes, I am the original offender .. I have sought forgiveness at the altar of Parsec and will continue to do so til my dying day)

File: cluster/kubernetes/policies.go (hunk 1 above)
Comment (squaremo) [inline cluster/kubernetes/policies.go]: (Can be tidied later)

Comment (paulbellamy) [inline cluster/kubernetes/policies.go]: Probably yeah.

**Pull request merged**